File: php/src/Algorithms/InsertionSort.php
<?php

declare(strict_types=1);

namespace DSA\Algorithms;

/**
 * Insertion sort algorithm.
 * Time: O(n^2) worst/average, O(n) best (already sorted)
 * Space: O(1)
 * Stable: Yes
 */
class InsertionSort
{
    /**
     * Sort an array in-place using insertion sort.
     *
     * @param array &$arr Array to sort (modified in-place)
     * @param callable|null $comparator Custom comparison function
     */
    public static function sort(array &$arr, ?callable $comparator = null): void
    {
        $comparator = $comparator ?? fn($a, $b) => $a <=> $b;
        $n = count($arr);

        for ($i = 1; $i < $n; $i++) {
            $key = $arr[$i];
            $j = $i - 1;

            while ($j >= 0 && $comparator($arr[$j], $key) > 0) {
                $arr[$j + 1] = $arr[$j];
                $j--;
            }
            $arr[$j + 1] = $key;
        }
    }

    /**
     * Return a sorted copy of the array using insertion sort.
     *
     * @param array $arr Array to sort
     * @param callable|null $comparator Custom comparison function
     * @return array Sorted array
     */
    public static function sorted(array $arr, ?callable $comparator = null): array
    {
        self::sort($arr, $comparator);
        return $arr;
    }
}

File: ruby/lib/algorithms/insertion_sort.rb
# frozen_string_literal: true

module DSA
  module Algorithms
    # Insertion sort algorithm.
    # Time: O(n²) worst/average, O(n) best (already sorted)
    # Space: O(1)
    # Stable: Yes
    module InsertionSort
      class << self
        # Sort an array in-place.
        def sort!(arr, &comparator)
          comparator ||= ->(a, b) { a <=> b }

          (1...arr.size).each do |i|
            key = arr[i]
            j = i - 1

            while j >= 0 && comparator.call(arr[j], key) > 0
              arr[j + 1] = arr[j]
              j -= 1
            end
            arr[j + 1] = key
          end

          arr
        end

        # Return a sorted copy of the array.
        def sort(arr, &comparator)
          sort!(arr.dup, &comparator)
        end
      end
    end
  end
end
